Prompt user to overwrite existing files

Sweetman opened the pull request:
### What does this PR do?
Addresses bug where files will be overwritten without a prompt. Add postcondition checker
### What issues does this PR fix or reference?
@W-4252622@

Sweetman commented on packages/salesforcedx-vscode-core/src/commands/commands.ts:
```diff
@@ -24,6 +24,70 @@ export interface PreconditionChecker {
   check(): boolean;
 }
 
+export interface PostconditionChecker<T> {
```
Added a PostconditionChecker interface that will check inputs gathered by ParameterGatherers. If it passes then the inputs will just be passed along, otherwise it will be replaced by a CancelResponse.

ruthemmanuelle commented on packages/salesforcedx-vscode-core/src/messages/i18n.ts:
```diff
@@ -70,5 +70,9 @@ export const messages = {
-  force_lightning_interface_create_text: 'SFDX: Create Lightning Interface'
+  force_lightning_interface_create_text: 'SFDX: Create Lightning Interface',
+  warning_prompt_file_override:
+    'File(s) with specified path already exist in workspace. Would you like to override?',
```
Should "override" be "overwrite"?

Sweetman replied:
Right, good catch. Just started to notice that I've been mentally using them interchangeably which is incorrect. I'll update.

Sweetman commented on packages/salesforcedx-vscode-core/src/commands/commands.ts:
```diff
@@ -24,6 +24,70 @@ export interface PreconditionChecker {
+  ): Promise<ContinueResponse<T> | CancelResponse>;
+}
+
+export class FilePathExistsChecker
```
This will check if the file path given by the user already contains the file name they specified. If it does, then a warning prompt will show asking to cancel will show. Yes will pass the ContinueResponse along, No will replace with CancelResponse

vazexqi replied:
It seems like it would better to create two postcondition checkers: one for the general case and one for the Lightning case. That way, it's easier to understand what is going on instead of having to read this `FilePathExistsChecker` and figure out why there is an 'if' statement case.

Then for the Lightning commands, those will specifically use the Lightning-specific one.

Sweetman commented on packages/salesforcedx-vscode-core/src/commands/commands.ts:
```diff
@@ -24,6 +24,70 @@ export interface PreconditionChecker {
+  }
+}
+
+export class EmptyPostChecker implements PostconditionChecker<any> {
```
Added an EmptyPostChecker that will be the default for already existing commands.

Sweetman commented on packages/salesforcedx-vscode-core/src/commands/commands.ts:
```diff
@@ -219,24 +283,27 @@ export abstract class SfdxCommandletExecutor<T>
     gatherer: ParametersGatherer<T>,
-    executor: CommandletExecutor<T>
+    executor: CommandletExecutor<T>,
+    postchecker = new EmptyPostChecker()
```
Make PostconditionChecker optional by specifying the default EmptyPostChecker.

Sweetman commented on packages/salesforcedx-vscode-core/src/messages/i18n.ts:
```diff
@@ -70,5 +70,9 @@ export const messages = {
   force_lightning_event_create_text: 'SFDX: Create Lightning Event',
-  force_lightning_interface_create_text: 'SFDX: Create Lightning Interface'
+  force_lightning_interface_create_text: 'SFDX: Create Lightning Interface',
+  warning_prompt_file_override:
```
@ruthemmanuelle Could you take a look at this message? The context is what to prompt when the user tries to create an Apex/Lightning/VF file that already exists. I made it 'file(s)' because in Lightning's case, multiple files will be overwritten.

Sweetman commented:
@vazexqi I would like some feedback on this before I add it to the other create commands. We briefly talked about adding a PostconditionChecker when we talked about the Commandlet framework so I want to make sure this is the direction you had in mind.

vazexqi commented on packages/salesforcedx-vscode-core/src/commands/commands.ts:
```diff
@@ -24,6 +24,70 @@ export interface PreconditionChecker {
+          nls.localize('warning_prompt_yes'),
+          nls.localize('warning_prompt_no')
+        );
+        if (override === 'Yes') {
```
I'm not sure if this will always be 'Yes'. 
If you call nls.localize on the message before showing it, won't it be whatever the localized message is?

vazexqi commented on packages/salesforcedx-vscode-core/src/commands/commands.ts:
```diff
@@ -24,6 +24,70 @@ export interface PreconditionChecker {
+  public constructor(fileExtension: string) {
+    this.fileExtension = fileExtension;
+  }
+  public getGlobPattern(
```
Formatting. Add a newline between methods.

Sweetman replied:
Method was deleted when separated Lightning into a separate PostconditionChecker

Sweetman commented on packages/salesforcedx-vscode-core/src/messages/i18n.ts:
```diff
@@ -70,5 +70,11 @@ export const messages = {
+  force_lightning_interface_create_text: 'SFDX: Create Lightning Interface',
+  warning_prompt_file_overwrite:
+    'File(s) with specified path already exists in workspace. Would you like to overwrite?',
+  warning_prompt_lightning_bundle_overwrite:
```
@ruthemmanuelle Could you check this new message? Context is that Lightning is a little different than the other create commands since it will place the files into a new directory with the name specified by the user.

We want to warn the user of 2 things with the message:
1) This directory is already in use by a lightning bundle (event, cmp, app, interface)
2) Depending on the bundle being created, some files may be overwritten since there are overlap in the files being created. See attached screenshot for an example.

![image](https://user-images.githubusercontent.com/4503716/30300396-4a2af980-9709-11e7-8739-0fc228d64c9d.png)

/cc @vazexqi

ruthemmanuelle commented on packages/salesforcedx-vscode-core/src/messages/i18n.ts:
```diff
@@ -72,5 +72,11 @@ export const messages = {
-  force_source_status_remote_text: 'SFDX: View Changes in Default Scratch Org'
+  force_source_status_remote_text: 'SFDX: View Changes in Default Scratch Org',
+  warning_prompt_file_overwrite:
+    'File(s) with specified path already exists in workspace. Would you like to overwrite?',
```
'One or more files with the specified path already exist in your workspace. Do you want to overwrite them?'

Sweetman commented:
@vazexqi I have added the integration tests for the PostconditionCheckers and removed the (potentially flaky) spectron tests. I had to add a directory into the system-tests assets folder containing a lightning bundle for the integration tests.

One of the previous tests that verifies a glob expression depends on the number of directories in the assets folder so I am considering making the core package have it's own workspace. Seems somewhat odd for a test in core to be dependent on the spectron tests assets. Thoughts?

Other than that, I think it's ready to be merged.
